### PR TITLE
Some improvements to ZCohomology

### PR DESCRIPTION
**Pull request by aljungstrom**
This PR contains quite a lot of stuff, but I think all changes in `Homotopy` and `ZCohomology` can be ignored. The same goes for `FromNegOne`.

All changes are made in order to get the cohomology groups to compute better. The main changes are:

- Redefining n-spheres as suspensions of S1. This removes a lot of annoying translations between the two definitions.
- Using the `FromNegOne` truncation in the definitions of connectedness and cohomology in order to avoid the additional constructors in the usual definition of truncations.
- Avoiding, to an extent as great as possible, transports into univalence in `HLevels`.
- Removing eta from `Iso`

These changes also let me remove most of the abstract in the concerned modules.

After these changes, both the 0th and the 1st cohomology groups compute decently. This confirms my suspicion that Freudenthal is the main culprit (only used for n>1) here. If anyone has a more direct proof of the fact that 
Ω ∥ Sⁿ⁺¹ ∥ₙ₊₁ ≃ ∥ Sⁿ ∥ₙ (especially for n = 2), please tell me:-)

### CODE REVIEW

File: Cubical/Algebra/Group/Algebra.agda
Comment (mortberg) [inline Cubical/Algebra/Group/Algebra.agda:250]: This name is really bad. Please rename or make private where it's used

File: Cubical/Data/Sigma/Properties.agda
Comment (mortberg) [inline Cubical/Data/Sigma/Properties.agda:81]: No parentheses needed around first equality argument

File: Cubical/Data/Sigma/Properties.agda
Comment (mortberg) [inline Cubical/Data/Sigma/Properties.agda:213]: Commented code, please delete

File: Cubical/Data/Unit/Base.agda
Comment (mortberg) [inline Cubical/Data/Unit/Base.agda:3]: Do you need the whole Prelude or just Core?

File: Cubical/Data/Unit/Base.agda
Comment (mortberg) [inline Cubical/Data/Unit/Base.agda:10]: Maybe it would be better if this would be the default Unit? Also the commented is strange, why not "universe polymorphic version"?

File: Cubical/Data/Unit/Properties.agda
Comment (mortberg) [inline Cubical/Data/Unit/Properties.agda:58]: ---?

File: Cubical/Data/Unit/Properties.agda
Comment (mortberg) [inline Cubical/Data/Unit/Properties.agda:59]: Do you really need to parametrize over \ell here?

Comment (mortberg) [reply Cubical/Data/Unit/Properties.agda:59]: Same for the lemmas below

File: Cubical/Foundations/Equiv/HalfAdjoint.agda
Comment (mortberg) [inline Cubical/Foundations/Equiv/HalfAdjoint.agda:153]: Yikes, this had a scary proof! Can't one do it by `elimIso` (https://github.com/agda/cubical/blob/master/Cubical/Foundations/Univalence.agda#L287) instead?

I guess one potential problem with that lemma is that it's using univalence... But maybe that's still simpler this proof? (which might be using univalence somewhere anyway...)

File: Cubical/Foundations/HLevels.agda
Comment (mortberg) [inline Cubical/Foundations/HLevels.agda:186]: This type is really strangely indendented. Same for the other similar lemmas

File: Cubical/Foundations/HLevels.agda
Comment (mortberg) [inline Cubical/Foundations/HLevels.agda:226]: Hmm, wouldn't this be easier to read if you inline the uses of cong?

File: Cubical/Foundations/HLevels.agda
Comment (mortberg) [inline Cubical/Foundations/HLevels.agda:284]: I would call this `section-Σ≡Prop`

File: Cubical/Foundations/HLevels.agda
Comment (mortberg) [inline Cubical/Foundations/HLevels.agda:293]: I would call this `isEquiv-Σ≡Prop`

File: Cubical/Functions/Fibration.agda
Comment (mortberg) [inline Cubical/Functions/Fibration.agda:83]: Commented code

File: Cubical/HITs/Hopf.agda
Comment (mortberg) [inline Cubical/HITs/Hopf.agda:326]: What is the other use on `inv` in scope? Maybe that should be renamed?

File: Cubical/HITs/Join/Base.agda
Comment (mortberg) [inline Cubical/HITs/Join/Base.agda:87]: Redundant parantheses

File: Cubical/HITs/S2/Base.agda
Comment (mortberg) [inline Cubical/HITs/S2/Base.agda:10]: Move import to the top

File: Cubical/HITs/S2/Base.agda
Comment (mortberg) [inline Cubical/HITs/S2/Base.agda:11]: Redundant parentheses in type and not necessary to pattern-match on {A = A}

File: Cubical/HITs/Sn/Base.agda
Comment (mortberg) [inline Cubical/HITs/Sn/Base.agda:18]: I think you can write 0 and 1 instead of `zero` and `suc zero`

File: Cubical/HITs/Truncation/FromNegOne/Base.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Base.agda:19]: I would not introduce the ^-1 here. Maybe better to just call it `HubAndSpoke` or something?

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:162]: This should be moved to theory about naturals file

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:157]: This lemma is a bit strange... It's really just computing `n - min n m`

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:172]: Long line

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:174]: Move import up

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:181]: Long line

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:242]: Do you really need elim in the fun/inv cases and not just rec? Same for the other similar lemmas

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:279]: Naming is weird

File: Cubical/HITs/Truncation/FromNegOne/Properties.agda
Comment (mortberg) [inline Cubical/HITs/Truncation/FromNegOne/Properties.agda:285]: Why not use the lemma above?

File: Cubical/Homotopy/Connected.agda
Comment (mortberg) [inline Cubical/Homotopy/Connected.agda:183]: It seems like this proof pattern is appearing a lot. You apply "isContrRetract" (or similar lemmas) so some iso which you need to decompose. Why not prove this once and for all as a lemma "isContrRetractFromIso" or something?

File: Cubical/Homotopy/Freudenthal.agda
Comment (mortberg) [inline Cubical/Homotopy/Freudenthal.agda:97]: redundant parentheses around lem

File: Cubical/Homotopy/Freudenthal.agda
Comment (mortberg) [inline Cubical/Homotopy/Freudenthal.agda:125]: longline

File: Cubical/Homotopy/Freudenthal.agda
Comment (mortberg) [inline Cubical/Homotopy/Freudenthal.agda:121]: move imports up

**Review by mortberg — CHANGES_REQUESTED** [review]
Very nice that things are faster! I commented on everything except in the ZCohomology files as suggested.

Looking at this code makes me think that FromNegOne should be the default Truncation while the current one should be in a separate folder?

File: Cubical/Data/Unit/Base.agda
Comment (ecavallo) [reply Cubical/Data/Unit/Base.agda:10]: Might be better to define it using the existing `Lift` and `Unit`:
```
Unit* : {ℓ : Level} → Type ℓ
Unit* = Lift Unit

pattern tt* = lift tt
```
I suspect that this would be inconvenient to use as the default `Unit`. For one thing, you're often going to need to specify the level explicitly, `Unit* {ℓ}`, and you'll need more level quantification everywhere.

Comment (mortberg) [reply Cubical/Data/Unit/Base.agda:10]: Ah ok, I agree with @ecavallo 's suggestion then!

File: Cubical/Algebra/Group/Algebra.agda
Comment (aljungstrom) [reply Cubical/Algebra/Group/Algebra.agda:250]: Haha wups, didn't mean to keep that name. I think it's a pretty useful lemma, so it should probably be there. But I'll rename it.

File: Cubical/Data/Unit/Base.agda
Comment (aljungstrom) [reply Cubical/Data/Unit/Base.agda:10]: Right, the lifted one is probably better to use with truncations since it inherits definitional equality from Unit. The reason I defined it like this is that I didn't want the definitional equality (I "lock" theorems in ZCohomology by giving them with respect to an element in Unit* and that doesn't work if you have definitional equality). I think I'll add a third Unit type called "lockUnit" or something then.

File: Cubical/Foundations/Equiv/HalfAdjoint.agda
Comment (aljungstrom) [reply Cubical/Foundations/Equiv/HalfAdjoint.agda:153]: Hmm, I'm a bit scared of this proof. But I realised it was super easy to prove with J. This doesn't even rely on the fact that we have an Iso, so that should be much better for computations.

File: Cubical/Data/Unit/Base.agda
Comment (Saizan) [reply Cubical/Data/Unit/Base.agda:10]: Others have used the name `Block` for a unit type used like this. Even giving it a phantom string parameter, `Block : String -> Type` so the `String` can document what you're blocking.

File: Cubical/HITs/Hopf.agda
Comment (aljungstrom) [reply Cubical/HITs/Hopf.agda:326]: That's the function from S1 to S1 sending loop to sym loop. I renamed it to invLooper

File: Cubical/Foundations/Equiv/HalfAdjoint.agda
Comment (mortberg) [reply Cubical/Foundations/Equiv/HalfAdjoint.agda:153]: I'm confused, what are you using J on to prove it? 

As far as I can tell the proof is still scary btw...

**Review by mortberg — APPROVED** [review]

**Comment by mortberg** [comment]
I would be happy to merge once the confusion about `invCongFunct` has been resolved

**Comment by mortberg** [comment]
After merging this the beginning of the output of `make timings` is:

```
Cubical.ZCohomology.MayerVietorisUnreduced       14,023ms 
Cubical.ZCohomology.Properties                   12,158ms 
Cubical.ZCohomology.Groups.Sn                     8,331ms 
Cubical.Data.DescendingList.Examples              7,971ms 
Cubical.ZCohomology.Groups.Torus                  6,908ms 
Cubical.Algebra.Algebra.Base                      6,465ms 
Cubical.ZCohomology.Groups.Prelims                5,337ms 
Cubical.Foundations.GroupoidLaws                  5,008ms 
Cubical.Data.Fin.Properties                       4,479ms 
Cubical.HITs.Hopf                                 4,454ms 
Cubical.Foundations.HLevels                       3,973ms 
Cubical.Foundations.Id                            3,888ms 
Cubical.HITs.Truncation.FromNegOne.Properties     3,477ms 
Cubical.Algebra.FreeCommAlgebra                   3,467ms 
Cubical.Algebra.CommAlgebra.Base                  3,349ms 
Cubical.Homotopy.Connected                        3,243ms 
Cubical.HITs.S1.Base                              3,131ms
```

After running `agda --ignore-interfaces -v profile.definitions:10 Cubical/README.agda` I get the following output:

```
Cubical.Data.DescendingList.Examples.LongerExample                                  9,430ms 
Cubical.ZCohomology.Properties.commₖ                                                3,369ms 
Cubical.ZCohomology.MayerVietorisUnreduced.MV.dHomHelperPath                        2,654ms 
Cubical.Foundations.Id._.f12                                                        2,360ms 
Cubical.HITs.KleinBottle.Properties._                                               2,071ms 
Cubical.Foundations.Pointed.Homotopy._.∙∼≃∙∼P                                       1,569ms 
Cubical.ZCohomology.Groups.Torus._.theIso                                           1,420ms 
Cubical.Homotopy.Freudenthal._.encodeMerid                                          1,328ms 
Cubical.ZCohomology.Groups.Torus.H¹-T²≅ℤ×ℤ                                          1,320ms 
Cubical.ZCohomology.Properties.rUnitlUnit0                                          1,293ms 
Cubical.HITs.Cost.Base._                                                            1,070ms 
Cubical.Homotopy.Freudenthal._.isEquivFwd                                             923ms 
Cubical.ZCohomology.Groups.Prelims._._._.helperFun2                                   868ms 
```

I'll make a PR commenting out the first of these and see if I can do anything about the other slow ones

**Pull request merged**
